sys/kern/uipc_shm.c

```diff
@@ -1490,45 +1504,59 @@ shm_mmap_large(struct shmfd *shmfd, vm_map_t map, vm_offset_t *addr,
 		if (align < pagesizes[shmfd->shm_lp_psind])
 			return (EINVAL);
 	}
+	align = MAX(align, CHERI_REPRESENTABLE_ALIGNMENT(size));
 
 	vm_map_lock(map);
 	if ((flags & MAP_FIXED) == 0) {
 		try = 1;
-		if (curmap && (*addr == 0 ||
-		    (*addr >= round_page((vm_offset_t)vms->vm_taddr) &&
-		    *addr < round_page((vm_offset_t)vms->vm_daddr +
+		if (curmap && (vaddr == 0 ||
+		    (vaddr >= round_page((vm_offset_t)vms->vm_taddr) &&
+		    vaddr < round_page((vm_offset_t)vms->vm_daddr +
 		    lim_max(td, RLIMIT_DATA))))) {
-			*addr = roundup2((vm_offset_t)vms->vm_daddr +
+			vaddr = roundup2((vm_offset_t)vms->vm_daddr +
 			    lim_max(td, RLIMIT_DATA),
 			    pagesizes[shmfd->shm_lp_psind]);
 		}
 again:
-		rv = vm_map_find_aligned(map, addr, size, maxaddr, align);
+		rv = vm_map_find_aligned(map, &vaddr, size, maxaddr, align);
 		if (rv != KERN_SUCCESS) {
 			if (try == 1) {
 				try = 2;
-				*addr = vm_map_min(map);
-				if ((*addr & mask) != 0)
-					*addr = (*addr + mask) & mask;
+				vaddr = vm_map_min(map);
+				if ((vaddr & mask) != 0)
+					vaddr = (vaddr + mask) & mask;
 				goto again;
 			}
 			goto fail1;
 		}
-	} else if ((flags & MAP_EXCL) == 0) {
-		rv = vm_map_delete(map, *addr, *addr + size);
+
+		reservation = vaddr;
+		rv = vm_map_reservation_create_locked(map, &reservation,
+		    size, max_prot);
 		if (rv != KERN_SUCCESS)
 			goto fail1;
+		rv = vm_map_insert(map, shmfd->shm_object, foff,
+		    reservation, reservation + size, prot, max_prot,
+		    docow, reservation);
 	} else {
-		error = ENOSPC;
-		if (vm_map_lookup_entry(map, *addr, &prev_entry))
-			goto fail;
-		next_entry = vm_map_entry_succ(prev_entry);
-		if (next_entry->start < *addr + size)
+		/* MAP_FIXED */
```

Comment:
The loss of a separate MAP_EXCL case concerns me; do we have kyua tests to cover the upstream behaviour that we can run to verify this doesn't break things?

Reply:
This has now changed. I found kyua tests that were not being ran that check this behaviour and fixed them. There is one that seems to be failing in dev as well and I have not debugged it thoroughly. The rest has been checked on amd64, I need to check on riscv as it has superpages but seemingly support for non-transparent large pages is disabled.